### PR TITLE
Enable parsing the local hosts database in lookups

### PR DESCRIPTION
**Pull request by ibauersachs**

### CODE REVIEW

**Comment by nresare** [comment]
This seems like a useful piece of functionality. I definitely think it makes sense to add. Depending on whether you feel ambitious or not, I think that it might make sense to re-structure the HostsFileParser for clarity (see inline comments). But this is clearly in the realm of opinion, and adding this I think improves the project so I'm happy to endorse either way.

File: src/main/java/org/xbill/DNS/hosts/HostsFileParser.java [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+    }
+
+    if (isEntireFileParsed || !Files.exists(path)) {
+      return Optional.empty();
+    }
+
+    return parseHostsFile(name, type);
+  }
+
+  private Optional<InetAddress> parseHostsFile(Name name, int type) throws IOException {
```
[the remark this answers is not in the document]
Comment (nresare) [reply src/main/java/org/xbill/DNS/hosts/HostsFileParser.java:108]: Introduce a new `Map<String, InetAddress> buildCache(Path)` that would simply build a new cache and replace the old one if it changed. For the case where we don't cache the whole file, the same parsing logic could be re-packaged in a `Optional<InetAddress> lookupAddress(Path, Name, int)`

**Review by nresare — APPROVED** [review]
Looking some more at the code, I realise that this would be a shared instance among Lookup instances potentially concurrently used from different threads, and the cache is a simple HashMap that potentially several of those threads would write to and read from at once. 

Concurrency errors are pesky to deal with because they typically only happens under certain circumstances. The simple fix would be to change it to a ConcurrentHashMap. With that change I am happy to approve.

**Comment by ibauersachs** [comment]
`getAddressForHost` is synchronized, so making `hostsCache` a `ConcurrentHashMap` doesn't make much sense. And the other way around (remove synchronized) could lead to parsing the hosts file over and over again in parallel.

I tried quite a bit to make the method less complicated, but I'll give it another round now that some time has passed.

File: src/main/java/org/xbill/DNS/hosts/HostsFileParser.java [hunk 2]
```diff
@@ -102,62 +103,89 @@ public HostsFileParser(Path path, boolean clearCacheOnChange) {
...
-            InetAddress lineAddress =
-                InetAddress.getByAddress(lineName.toString(true), lineAddressBytes);
-            InetAddress previous =
-                hostsCache.putIfAbsent(key(lineName, lineAddressType), lineAddress);
-            if (type == lineAddressType) {
-              if (!parseEntireFile) {
-                return Optional.of(lineAddress);
-              } else if (isSearchedEntry && previous == null) {
-                result = lineAddress;
+  private void searchHostsFileForEntry(Name name, int type) throws IOException {
```
Comment (nresare) [inline src/main/java/org/xbill/DNS/hosts/HostsFileParser.java:134]: This was pretty much exactly what I had in mind 😊

**Review by nresare — APPROVED** [review]

**Pull request merged**